src/org/opensolaris/opengrok/history/GitRepository.java

```diff
@@ -170,11 +125,48 @@ Executor getHistoryLogExecutor(final File file, String sinceRevision)
         cmd.add("--pretty=fuller");
         cmd.add("--date=rfc");
 
+        if (file.isFile() && RuntimeEnvironment.getInstance().isHandleHistoryOfRenamedFiles()) {
+            cmd.add("--follow");
+        }
+
+        if (sinceRevision != null) {
+            cmd.add(sinceRevision + "..");
+        }
+
+        if (filename.length() > 0) {
+            cmd.add(filename);
+        }
+
+        return new Executor(cmd, new File(getDirectoryName()), sinceRevision != null);
+    }
+
+    Executor getRenamedFilesExecutor(final File file, String sinceRevision) throws IOException {
+        String abs = file.getCanonicalPath();
+        String filename = "";
+        if (abs.length() > directoryName.length()) {
+            filename = abs.substring(directoryName.length() + 1);
+        }
+
+        ensureCommand(CMD_PROPERTY_KEY, CMD_FALLBACK);
+        List<String> cmd = new ArrayList<>();
+        cmd.add(RepoCommand);
+        cmd.add("log");
+        cmd.add("-M8"); // similarity 80%
+        cmd.add("--summary");
+        cmd.add(ABBREV_LOG);
+        cmd.add("--name-status");
+        cmd.add("--oneline");
+
+        if (file.isFile()) {
+            cmd.add("--follow");
+        }
+
         if (sinceRevision != null) {
             cmd.add(sinceRevision + "..");
         }
 
         if (filename.length() > 0) {
```

Comment:
I modified the lines commented in [this comment](https://github.com/OpenGrok/OpenGrok/pull/1345#discussion_r97753146) and moved them here.

It's not really check for a file but anything under the git repository (not the repository itself). The following git command will throw an error if the path does not exist.